### PR TITLE
Alistair/4706 dyanmic reminders about survey tasks

### PR DESCRIPTION
**Pull request by alistairclark**
Unit tests are in a separate PR: #208 

Need to have a discussion with Luke about when these reminders should be sent, and so when to schedule the task.

I should probably suppress sending of these on test/staging environments too, but I'd like to make sure they work first.

### CODE REVIEW

File: api/endpoints/common/sql_queries.py [hunk 1]
```diff
@@ -334,6 +334,42 @@ def subquery(query):
...
+    AND NOT EXISTS(
+        SELECT
+            *
+        FROM
+            public.projects_usertaskreminder utr
+            inner join public.projects_usertask ut2 on ut2.id = utr.user_task_id
+            inner join public.projects_userproject up2 on up2.id = ut2.user_project_id
+            inner join public.projects_user u2 on u2.id = up2.user_id
+        WHERE
+            utr.created >= '{datetime_since_reminder_sent}'
```
Comment (this-andy) [inline api/endpoints/common/sql_queries.py:369]: Can we ensure we don't send two inactivity reminders for same task

**Review by this-andy — CHANGES_REQUESTED** [review]

File: api/endpoints/common/sql_queries.py [hunk 2]
```diff
@@ -334,6 +334,44 @@ def subquery(query):
...
+    AND utr.id is null
+    AND NOT EXISTS(
+        SELECT
+            *
+        FROM
+            public.projects_usertaskreminder utr
+            inner join public.projects_usertask ut on ut.id = utr.user_task_id
+            inner join public.projects_userproject up on up.id = ut.user_project_id
+            inner join public.projects_user u on u.id = up.user_id
+        WHERE
```
Comment (this-andy) [inline api/endpoints/common/sql_queries.py:370]: This looks like it will result in multiple reminders being sent for an individual task - every HOURS_SINCE_LAST_REMINDER_SENT.  Am I reading it right (it's kind of difficult to be certain of the exact behaviour of SQL when it's complex like this).
I think it is ok to send multiple reminders for different tasks (as we discussed in meeting a few days ago), but not multiple reminders for a single task.

File: api/endpoints/user_task_reminder.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    Queries the DB for a list of users who have not logged in within
+    a given time frame, or recently received a reminder email, but have
+    an incomplete task in progress.
+
+    Args:
+        task_type (str): the name of the task type (e.g. "survey", "interview")
+        reminder_type (str): the type of reminder to check for the existence of
+            (e.g. "inactivity" or "ending")
+    """
+    datetime_since_reminder_sent = datetime.utcnow() - timedelta(
```
Comment (this-andy) [inline api/endpoints/user_task_reminder.py:113]: Can you remind me what this does?  What does it represent.  It may be that my confusion comes from the name - does it describe its purpose well?

**Review by this-andy — CHANGES_REQUESTED** [review]

File: api/endpoints/user_task_reminder.py (hunk 3 above)
Comment (alistairclark) [reply api/endpoints/user_task_reminder.py:113]: Good point, the name here really isn't clear. I'll change it. 

The variable refers to the window of time within which we're looking for any other reminders being sent.

Comment (alistairclark) [reply api/endpoints/user_task_reminder.py]: I've renamed the variable slightly and added some comments to explain what these variables do. If you've got any thoughts on a better name then let me know though.

File: api/endpoints/common/sql_queries.py (hunk 2 above)
Comment (alistairclark) [reply api/endpoints/common/sql_queries.py:370]: I've been trying to add as many scenarios in the tests as I can because I agree the query is hard to follow. There's a test that covers this case though.

https://github.com/THIS-Institute/thiscovery-core/pull/208/files#diff-30f1a72674d74c7258c82c771e3e687f02a78268d9b5d07b431265bc553e891dR124

Comment (this-andy) [reply api/endpoints/common/sql_queries.py:370]: Yes, I agree, that SQL looks like it will give the behaviour we want.  Thanks for clarifying

**Review by this-andy — APPROVED** [review]
